### ISSUE TITLE
💅 `lint/nursery/useIterableCallbackReturn` incorrectly flags functions that return `void` as returning a value

### ISSUE THREAD

**Issue by DayKev** [biomejs/biome#6617]
### Environment information

```bash
CLI:
  Version:                      2.0.5
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           windows

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         unset
  JS_RUNTIME_VERSION:           v22.16.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/10.12.2

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.jsonc
  Formatter enabled:            false
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  true
  Enabled rules:
    a11y/noAccessKey
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExcessiveCognitiveComplexity
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringRaw
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useSimpleNumberKeys
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noPrivateImports
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useIsNan
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    nursery/noAwaitInLoop
    nursery/noConstantBinaryExpression
    nursery/noShadow
    nursery/noTsIgnore
    nursery/useAdjacentGetterSetter
    nursery/useIndexOf
    nursery/useIterableCallbackReturn
    nursery/useObjectSpread
    performance/noAccumulatingSpread
    performance/noDelete
    performance/noDynamicNamespaceImportAccess
    performance/noNamespaceImport
    security/noBlankTarget
    security/noDangerouslySetInnerHtml
    security/noDangerouslySetInnerHtmlWithChildren
    security/noGlobalEval
    style/noCommonJs
    style/noDescendingSpecificity
    style/noRestrictedTypes
    style/noUnusedTemplateLiteral
    style/noUselessElse
    style/useArrayLiterals
    style/useAsConstAssertion
    style/useBlockStatements
    style/useConsistentArrayType
    style/useConst
    style/useDefaultParameterLast
    style/useDeprecatedReason
    style/useExportType
    style/useFilenamingConvention
    style/useImportType
    style/useLiteralEnumMembers
    style/useNodejsImportProtocol
    style/useNumberNamespace
    style/useShorthandFunctionType
    suspicious/noApproximativeNumericConstant
    suspicious/noArrayIndexKey
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConstEnum
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noOctalEscape
    suspicious/noPrototypeBuiltins
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noVar
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```

### Rule name

lint/nursery/useIterableCallbackReturn

### Playground link

https://biomejs.dev/playground/?lineWidth=120&indentStyle=space&lintRules=useIterableCallbackReturn&code=YwBvAG4AcwB0ACAAdABlAHMAdABBAHIAcgAgAD0AIABbADEALAAgADIALAAgADMAXQA7AAoAZgB1AG4AYwB0AGkAbwBuACAAdABlAHMAdABGAG4AKABwAGEAcgBhAG0AOgAgAG4AdQBtAGIAZQByACkAOgAgAHYAbwBpAGQAIAB7AAoAIAAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoAHAAYQByAGEAbQApADsACgB9AAoAdABlAHMAdABBAHIAcgAuAGYAbwByAEUAYQBjAGgAKAAoAHYAKQAgAD0APgAgAHQAZQBzAHQARgBuACgAdgApACkAOwA%3D

### Expected result

`x.forEach((v) => fn(v));` should not trigger this rule when the return type of `fn` is `void`.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by siketyan** [comment]
I don't think this is a bug, as we don't know whether the return type of the function is void or not without type inference.

**Comment by DayKev** [comment]
Hmm... that's going to make it too noisy imo. Can this rule be changed to [optionally?] use the new type inference system?

**Comment by pastelmind** [comment]
Alternatively, could we exempt `void` expressions?

```js
values.forEach((foo) => foo.doSomething());        // Lint error
values.forEach((foo) => void foo.doSomething());   // OK
```

Note that this is same as ESLint's [`array-callback-return`](https://eslint.org/docs/latest/rules/array-callback-return) rule with `allowVoid` enabled.

**Comment by ematipico** [comment]
> Hmm... that's going to make it too noisy imo. Can this rule be changed to [optionally?] use the new type inference system?

Not really, unfortunately

**Comment by lantua** [comment]
Could we at least auto-fix `() => expression` to `() => { expression }` since this is a breaking change in 2.1.4 => 2.2.0. That said, I don't like that the formatter forces `{ expression }` to be multiline in combination with this fix in this case.

**Comment by DayKev** [comment]
Yeah if you only have a single line in the `{}` (e.g. `someArray.forEach((value) => { value.doThing(); });`) it would be nice if the formatter didn't force that to be multi-line if it doesn't go over the max line width, then it wouldn't matter that the rule can't see when the function returns `void` because you can just wrap it in `{}` and have it still be on a single line; and that seems like it could have an "unsafe" fixer as well I'd think.

**Comment by JoshStrobl** [comment]
Yea if it doesn't go over the column width, I don't think it should be forcing newlines for `{}`. Make it an option, sure, but the change results in a perfectly readable short line below:

```
    if (a.id) return set.add(a.id);
```

Into:

```
  if (a.id) {
    set.add(a.id);
    return;
  }
```

I try to encourage clean, early returns in our codebase. This expansion makes it unnecessarily verbose. I would consider this a breaking change. Echoing @lantua in I don't like the formatter forcing the expression multi-line.

**Comment by Netail** [comment]
> > Hmm... that's going to make it too noisy imo. Can this rule be changed to [optionally?] use the new type inference system?
> 
> Not really, unfortunately

Bummer, bumped from 2.1.3 to 2.2.4 today and ran into 8 of these false positives

**Comment by khaled65442** [comment]
https://github.com/biomejs/biome/issues/6617
